File: services/cards-consultation/src/main/java/org/opfab/cards/consultation/model/LightCard.java
/* Copyright (c) 2018-2024, RTE (http://www.rte-france.com)
 * See AUTHORS.txt
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * SPDX-License-Identifier: MPL-2.0
 * This file is part of the OperatorFabric project.
 */

package org.opfab.cards.consultation.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;

import lombok.*;
import org.springframework.data.annotation.Transient;
import java.time.Instant;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class LightCard {

    private String uid;
    private String id;
    private String publisher;
    private String processVersion;
    private String process;
    private String processInstanceId;
    private String state;
    private Instant lttd;
    private Instant publishDate;
    private Instant startDate;
    private Instant endDate;
    private Instant expirationDate;
    private SeverityEnum severity;
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private List<String> tags;
    private I18n title;
    private I18n summary;
    private String titleTranslated;
    private String summaryTranslated;
    @JsonIgnore
    private List<String> usersAcks;
    @Transient
    private Boolean hasBeenAcknowledged;
    private List<String> entitiesAcks;
    private List<String> entityRecipients;
    private List<String> entityRecipientsForInformation;
    private List<String> userRecipients;
    private List<String> groupRecipients;
    private List<String> usersReads;
    @Transient
    private Boolean hasBeenRead;
    private String parentCardId;
    private String initialParentCardUid;
    private List<String> entitiesAllowedToRespond;
    private List<String> entitiesRequiredToRespond;
    private List<String> entitiesAllowedToEdit;
    private PublisherTypeEnum publisherType;
    private String representative;
    private PublisherTypeEnum representativeType;
    private String wktGeometry;
    private String wktProjection;
    private Integer secondsBeforeTimeSpanForReminder;
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonProperty("rRule") // if we don't use this annotation, the field will be serialized as "rrule"
    private RRule rRule;
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private List<CardActionEnum> actions;
    private List<TimeSpan> timeSpans;

    public static LightCard copy(Card other) {
        LightCardBuilder builder = builder()
                .uid(other.getUid())
                .id(other.getId())
                .parentCardId(other.getParentCardId())
                .initialParentCardUid(other.getInitialParentCardUid())
                .publisher(other.getPublisher())
                .process(other.getProcess())
                .processVersion(other.getProcessVersion())
                .state(other.getState())
                .processInstanceId(other.getProcessInstanceId())
                .lttd(other.getLttd())
                .startDate(other.getStartDate())
                .publishDate(other.getPublishDate())
                .endDate(other.getEndDate())
                .expirationDate(other.getExpirationDate())
                .severity(other.getSeverity())
                .tags(other.getTags())
                .title(other.getTitle())
                .summary(other.getSummary())
                .titleTranslated(other.getTitleTranslated())
                .summaryTranslated(other.getSummaryTranslated())
                .hasBeenAcknowledged(other.getHasBeenAcknowledged())
                .entitiesAcks(other.getEntitiesAcks())
                .entityRecipients(other.getEntityRecipients())
                .entityRecipientsForInformation(other.getEntityRecipientsForInformation())
                .userRecipients(other.getUserRecipients())
                .groupRecipients(other.getGroupRecipients())
                .hasBeenRead(other.getHasBeenRead())
                .entitiesAllowedToRespond(other.getEntitiesAllowedToRespond())
                .entitiesRequiredToRespond(other.getEntitiesRequiredToRespond())
                .publisherType(other.getPublisherType())
                .representative(other.getRepresentative())
                .representativeType(other.getRepresentativeType())
                .wktGeometry(other.getWktGeometry())
                .wktProjection(other.getWktProjection())
                .secondsBeforeTimeSpanForReminder(other.getSecondsBeforeTimeSpanForReminder())
                .rRule(other.getRRule())
                .actions(other.getActions())
                .timeSpans(other.getTimeSpans());

        return builder.build();

    }
}

File: ui/main/src/app/model/LightCard.ts
/* Copyright (c) 2018-2025, RTE (http://www.rte-france.com)
 * See AUTHORS.txt
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * SPDX-License-Identifier: MPL-2.0
 * This file is part of the OperatorFabric project.
 */

import {I18n} from 'app/model/I18n';
import {TimeSpan} from 'app/model/TimeSpan';
import {RRule} from 'app/model/RRule';
import {Severity} from './Severity';
import {CardAction} from './CardAction';
import {PublisherType} from './PublisherType';

export class LightCard {
    constructor(
        readonly uid: string,
        readonly id: string,
        readonly publisher: string,
        readonly processVersion: string,
        readonly publishDate: number,
        readonly startDate: number,
        readonly endDate: number,
        readonly expirationDate: number,
        readonly severity: Severity,
        readonly hasBeenAcknowledged: boolean = false,
        readonly hasBeenRead: boolean = false,
        public hasChildCardFromCurrentUserEntity: boolean = false,
        readonly processInstanceId?: string,
        readonly lttd?: number,
        readonly title?: I18n,
        readonly summary?: I18n,
        readonly titleTranslated?: string,
        readonly summaryTranslated?: string,
        readonly tags?: string[],
        readonly timeSpans?: TimeSpan[],
        readonly rRule?: RRule,
        readonly process?: string,
        readonly state?: string,
        readonly parentCardId?: string,
        readonly initialParentCardUid?: string,
        readonly representative?: string,
        readonly representativeType?: PublisherType | string,
        readonly wktGeometry?: string,
        readonly wktProjection?: string,
        readonly entitiesAcks?: string[],
        readonly entityRecipients?: string[],
        readonly entityRecipientsForInformation?: string[],
        readonly entitiesAllowedToRespond?: string[],
        readonly entitiesRequiredToRespond?: string[],
        readonly entitiesAllowedToEdit?: string[],
        readonly publisherType?: PublisherType | string,
        readonly secondsBeforeTimeSpanForReminder?: number,
        readonly actions?: CardAction[]
    ) {}
}
